Fix path could be nil in test_puma_server tests

JuanitoFatas opened the pull request:
### Description

When I ran the test `bundle exec ruby test/test_puma_server.rb` locally with `ruby 3.1.0p0 (2021-12-25 revision fb4df44d16) [arm64-darwin21]` on macOS, I got two failing tests which complains path is `nil`<sup>[1]</sup>.

With the rewrite the use of `Tempfile`, now it works for my local dev env `for i in $(seq 1 10); do bundle exec ruby test/test_puma_server.rb; done` and CI seems happy, too. 



<details>
<summary>For references docs on <code>Tempfile</code> and <code>IO</code></summary>

- Ruby 2.4: [Tempfile](https://rubyapi.org/2.4/o/tempfile), [IO#binmode](https://rubyapi.org/2.4/o/io#method-i-binmode) and [IO#write](https://rubyapi.org/2.4/o/io#method-i-write)
- Ruby 2.5: [Tempfile](https://rubyapi.org/2.5/o/tempfile), [IO#binmode](https://rubyapi.org/2.5/o/io#method-i-binmode) and [IO#write](https://rubyapi.org/2.5/o/io#method-i-write)
- Ruby 2.6: [Tempfile](https://rubyapi.org/2.6/o/tempfile), [IO#binmode](https://rubyapi.org/2.6/o/io#method-i-binmode) and [IO#write](https://rubyapi.org/2.6/o/io#method-i-write)
- Ruby 2.7: [Tempfile](https://rubyapi.org/2.7/o/tempfile), [IO#binmode](https://rubyapi.org/2.7/o/io#method-i-binmode) and [IO#write](https://rubyapi.org/2.7/o/io#method-i-write)
- Ruby 3.0: [Tempfile](https://rubyapi.org/3.0/o/tempfile), [IO#binmode](https://rubyapi.org/3.0/o/io#method-i-binmode) and [IO#write](https://rubyapi.org/3.0/o/io#method-i-write)
- Ruby 3.1: [Tempfile](https://rubyapi.org/3.1/o/tempfile), [IO#binmode](https://rubyapi.org/3.1/o/io#method-i-binmode) and [IO#write](https://rubyapi.org/3.1/o/io#method-i-write)

</details>


<details>
<summary><sup>[1] Error I got</sup></summary>

```
  1) Error:
TestPumaServer#test_file_body:
TypeError: no implicit conversion of nil into String
    test/test_puma_server.rb:158:in `exist?'
    test/test_puma_server.rb:158:in `ensure in test_file_body'
    test/test_puma_server.rb:158:in `test_file_body'
    /Users/hhh/dev/puma/test/helper.rb:89:in `block (4 levels) in run'
    /Users/hhh/.rubies/ruby-3.1.0/lib/ruby/3.1.0/timeout.rb:107:in `block in timeout'
    /Users/hhh/.rubies/ruby-3.1.0/lib/ruby/3.1.0/timeout.rb:117:in `timeout'
    /Users/hhh/dev/puma/test/helper.rb:87:in `block (3 levels) in run'

  2) Error:
TestPumaServer#test_file_to_path:
TypeError: no implicit conversion of nil into String
    test/test_puma_server.rb:177:in `exist?'
    test/test_puma_server.rb:177:in `ensure in test_file_to_path'
    test/test_puma_server.rb:177:in `test_file_to_path'
    /Users/hhh/dev/puma/test/helper.rb:89:in `block (4 levels) in run'
    /Users/hhh/.rubies/ruby-3.1.0/lib/ruby/3.1.0/timeout.rb:107:in `block in timeout'
    /Users/hhh/.rubies/ruby-3.1.0/lib/ruby/3.1.0/timeout.rb:117:in `timeout'
    /Users/hhh/dev/puma/test/helper.rb:87:in `block (3 levels) in run'
```
</details>

### Your checklist for this pull request
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [ ] I have added (or updated) appropriate tests if this PR fixes a bug or adds a feature.
- [x] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [ ] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [ ] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed, including Rubocop.

dentarg commented:
What seed did the tests run with?

> somehow `Tempfile.open { |f| f.path } returns nil`

If that happened, shouldn't those tests have failed? Very strange indeed.

MSP-Greg commented:
> on macOS, I got two failing tests which complains path is nil (somehow `Tempfile.open { |f| f.path }` returns nil???)

Ok...  Good to know.

If you replace:
```ruby
    path = Tempfile.open { |f| f.path }
    File.binwrite path, random_bytes
```
with:
```ruby
    path = Tempfile.open(mode: File::BINARY) { |f|
      f.syswrite random_bytes
      f.path
    }
```

Does it work on macOS?

MSP-Greg commented:
@JuanitoFatas Forget about the last post.  Some of the method signatures for Tempfile vary from Ruby 2.4 to 3.1/master...

JuanitoFatas commented:
@dentarg @MSP-Greg I rewrote the test (trial and error...sorry I have no good theory of this version is better than the original) and updated Pull Request description, hopefully this works better for all environments 😅

MSP-Greg commented:
@JuanitoFatas

> trial and error...sorry I have no good theory

Although I think most people would want to know the why, but sometimes:

* Local errors not happening in CI
* CI failing but local is ok
* Some platforms fail (or worse, they fail intermittently), but others always pass locally and in CI

I worked a little bit on this yesterday.  In the last run, JRuby was failing, but MRI seemed ok.

Re Windows, it's a good idea to always close fd's, as Windows often doesn't like having a file opened more than once.

I think it might be a good idea to have a 'write a string to a tempfile and return the path' shared method.

Some tests load config files, and often they are very small.  At present, they are stored as files, which is a PITA.  If the tests had the config as a here doc, it's a lot easier/faster to see what the test is doing.  Make sense?

MSP-Greg commented:
> Which tests load small config files?

Not sure about which tests, but https://github.com/puma/puma/tree/master/test/config has a lot of very small files.

Also, I forgot another very good reason for this.  When the config is a pre-existing file, one can't do string interpolation on it, so we have a few tests that use predetermined file names.  When the config is a here doc in the actual test code, we can do string interpolation (using the tempfile path) and save the heredoc to a tempfile.

dentarg commented:
> Some tests load config files, and often they are very small. At present, they are stored as files, which is a PITA. If the tests had the config as a here doc, it's a lot easier/faster to see what the test is doing. Make sense?

Is it because you need to open the config file when reading the code for the test to get the full picture? Or would we get better backtraces when tests failed?

> https://github.com/puma/puma/tree/master/test/config 

I like that these files exist, I find them useful when testing things manually with Puma

MSP-Greg commented:
> Is it because you need to open the config file when reading the code for the test to get the full picture?

Yes, that along with the conf can be dynamic with string interpolation.

> I like that these files exist, I find them useful when testing things manually with Puma

Makes sense.

dentarg commented:
> the conf can be dynamic with string interpolation

That is a good argument

nateberkopec commented:
Windows test failures are weird. Any thoughts @MSP-Greg ?

MSP-Greg commented:
Re Windows, please see https://github.com/MSP-Greg/puma/commit/021237548bf2.

@JuanitoFatas Thoughts?  Using `yield` might be problematic...

This PR shows that some work needs to be done on the test system.  Several months ago I started on a rewrite, the main concept was improving the helper files, which I divided into three types:

1. Client and shared helper code - contains code for creating client sockets (either raw sockets or net/http) and also streams of clients.  Shares variables with the next two helpers, allowing easy changes between socket types (tcp, unix, ssl).  Cleanup code would be included to make sure all io's are closed. This would also have helper code for things like temp files.

2. Code for working with in-process servers which are started via `Server.new` / `Server#run`

3. Code for using out-of-process servers using `IO.popen`, like current integration tests.

Hopefully, I can review my old code and submit PR's with it soon.  We could then begin moving tests into new files, and maybe splitting them into more files, as some of the test files have a very large number of tests.

MSP-Greg commented:
@JuanitoFatas 

Does your local test with the added commit?

Sorry for the failing tests, as I used a smaller body.  The problem with tests is that the integration tests take time, as they're starting another process, but the tests like TestPumaServer can have odd blocking going on with multiple servers and clients.  Using normal writes/reads both Ubuntu & Windows took 20 seconds for the two tests...

JuanitoFatas commented:
> Does your local test pass with the added commit?

Yes [f9fac67](https://github.com/puma/puma/pull/2981/commits/f9fac67e8cb91ed5d2de4e801ee90b381dc90d70) passed for me 🎉

nateberkopec commented:
🙇‍♂️
